New module: Support manipulating XML files [files/xml]

dagwieers opened the pull request:
##### SUMMARY
This is another attempt to get the xml module upstream.
https://github.com/cmprescott/ansible-xml/

Prior submissions at: #16147, ansible/ansible-modules-core#4112 and ansible/ansible-modules-core#4122

This PR includes:
- The code has been re-licensed to GPLv3+. cmprescott/ansible-xml#113
- 55 integration tests (tested in run-mode and check-mode)
- Add return values to documentation
- Added python 3 support (revised unicode support)
- Improved and simplified check-mode support
- Fixed one idempotency bug
- Fixed a bug related to check-mode support (incorrectl reported no change)

Once merged we can collaborate on the module and tests in this project:
- Migrate and fix [open issues](https://github.com/cmprescott/ansible-xml/issues/)
- Merge [existing PRs](https://github.com/cmprescott/ansible-xml/pulls/)
- Add diff support !!  --  *Ideal in combination with check-mode*
- Add backup support

This fixes cmprescott/ansible-xml#52,
and this fixes cmprescott/ansible-xml#74,
and this fixes cmprescott/ansible-xml#75,
but his also fixes cmprescott/ansible-xml#108,
and this fixes cmprescott/ansible-xml#109,
and this fixes cmprescott/ansible-xml#116,

Due to the addition of a namespace example and a documentation note (as stated in cmprescott/ansible-xml#118),
this fixes cmprescott/ansible-xml#102,
and it fixes cmprescott/ansible-xml#107,
and fixes cmprescott/ansible-xml#112,
and also fixes cmprescott/ansible-xml#117,
and fixes cmprescott/ansible-xml#118.

Due to an improved lxml version check,
this fixes cmprescott/ansible-xml#114
and fixes cmprescott/ansible-xml#115.,

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
xml

##### ANSIBLE VERSION
v2.4

dagwieers committed:
Import original unmodified upstream version

This is another attempt to get the xml module upstream.
https://github.com/cmprescott/ansible-xml/

This is the original file from upstream,
without commit 1e7a3f6b6e2bc01aa9cebfd80ac5cd4555032774

dagwieers commented:
@abadger I would appreciate if you could look at how this module currently handles unicode (also wrt. Python 3) and bring this in line with what we nowadays expect for Ansible modules.

dagwieers committed:
Add additional changes required for upstreaming

This PR includes the following changes:

- Clean up of DOCUMENTATION
- Rename "ensure" parameter to "state" parameter (kept alias)
- Added EXAMPLES
- Remove explicit type-case using str() for formatting
- Clean up AnsibleModule parameter handling
- Retained Python 2.4 compatibility
- PEP8 compliancy
- Various fixes as suggested by abadger during first review

This fixes cmprescott/ansible-xml#108

dagwieers commented:
@gundalow @mattclay I have the integration tests ready, but they fail on docker lacking lxml. Installing it via pip in the playbook is not possible as it requires compilation, and adding it to the Dockerfile does not seem to impact the used images.

dagwieers committed:
Added original integration tests

There is som room for improvement wrt. idempotency and check-mode
testing.

bbaassssiiee commented:
This module would be a great addition creating capabilities for deploying entersprise (java) applications.

abadger commented:
I think there will be quite a few problems with it wrt bytes/unicode and python3.

The general rules are:
* You never want to mix bytes and text types. 
* You want to be explicit in whether you're dealing with bytes or text whenever dealing with io.

I see that according to comments in the code, lxml needs to use text strings too.  That probably means that the easiest way to code this module to be unicode safe is to transform everything into text (as opposed to turning it into native strings as is done in other modules).  Using to_text is the best way to do this.  the decode function that I see i nthis module could probably be done via ansible.module_utils.basic.json_dict_bytes_to_unicode instead but that function needs to move somewhere else in the future so it's not a requirement that the code use it.

dagwieers committed:
Some tests depend on lxml v3.0alpha1 or higher

We are now expecting lxml v2.3.0 or higher.
We skips tests if lxml is too old.

Plus small fix.

mattclay commented:
@dagwieers For the lxml requirement, add tasks to the start of the test:

- For FreeBSD, use the pkgng module to install the py27-lxml package.
- Install lxml using the pip module. That will take care of macOS.

The pip install will no-op if lxml is already installed. This way we only install it when needed.

With those changes the tests should pass CI.

dagwieers commented:
@mattclay Thanks, this is looking good !

richardcs commented:
Just made the [PR](https://github.com/ansible/ansible/pull/26404) for a windows version of things that does XML files on windows. @jhawkesworth provided a lot of input and the interface is the same with aliases. I tend to think the argument for the xml to be added should be 'fragment' since that's a well understood concept in the XML community. It does feel a little strange to propagate details about XML up to ansible, but using XPath is pretty heady stuff too.

dagwieers commented:
Every contributor has agreed to re-licensing this code, so this is ready to be merged.

dagwieers committed:
Relicense to GPLv3+ header

All past contributors have agreed to relicense this module to GPLv2+, and GPLv3 specifically.

See: https://github.com/cmprescott/ansible-xml/issues/113

This fixes cmprescott/ansible-xml#73

dagwieers commented:
@mattclay I assume the pylint tests are run for Python 3, whereas the xml module at this time is Python 2 only and would need some special love to make it work with Python 3. ~(I am not the guy to do that though)~ Python 3 now !

mattclay commented:
@dagwieers Yes, pylint is run using Python 3 on Shippable. All new modules need to support both python 2.x and 3.x.

/cc @abadger

dagwieers committed:
Fix small typo in integration tests

dagwieers committed:
Python 3 support

This PR also includes:
- Python 3 support
- Documentation fixes
- Check-mode fixes and improvements
- Bugfix in check-mode support
- Always return xmlstring, even if there's no change
- Check for lxml 2.3.0 or newer

resmo commented:
shipit

dagwieers committed:
Add return values

abadger commented on lib/ansible/modules/files/xml.py:
```diff
@@ -0,0 +1,757 @@
+    if xml_string:
+        infile = BytesIO(xml_string.encode('utf-8'))
+    elif os.path.isfile(xml_file):
+        infile = open(xml_file, 'r')
```
This needs to be 'rb' so that it's bytes just like the BytesIO above.

abadger commented on lib/ansible/modules/files/xml.py:
```diff
@@ -0,0 +1,757 @@
+    # Check if we have lxml 2.3.0 or newer installed
+    if not HAS_LXML:
+        module.fail_json(msg='The xml ansible module requires the lxml python library installed on the managed machine')
+    elif LooseVersion('.'.join(map(str, etree.LXML_VERSION))) < LooseVersion('2.3.0'):
```
Not necessary but this map and the one on 690 are better written as:
``` python
elif LooseVersion('.'.join(to_native(ver_field) for ver_field in etree.LXML_VERSION)) < LooseVersion('2.3.0'):

dagwieers replied:
Would that be the same as:
```python
elif LooseVersion('.'.join(map(to_native, etree.LXML_VERSION))) < LooseVersion('2.3.0'):
```
i.e. simply replacing str() with to_native().

abadger commented on lib/ansible/modules/files/xml.py:
```diff
@@ -0,0 +1,757 @@
+    # Try to parse in the target XML file
+    try:
+        parser = etree.XMLParser(remove_blank_text=pretty_print)
+        x = etree.parse(infile, parser)
```
Maybe document, doc, or root would be a better name for this variable?

abadger commented on lib/ansible/modules/files/xml.py:
```diff
@@ -0,0 +1,757 @@
+        x = etree.parse(infile, parser)
+    except etree.XMLSyntaxError:
+        e = get_exception()
+        module.fail_json(msg="Error while parsing path: %s" % e)
```
Not necessary but exception handling (since this is ansible-2.4+) cna be written as:
```
    except etree.XMLSyntaxError as e:
        module.fail_json(msg="Error while parsing path: %s" % to_native(e))
```

abadger commented on lib/ansible/modules/files/xml.py:
```diff
@@ -0,0 +1,757 @@
+    return [child_to_element(module, child, type) for child in children]
+
+
+def finish(module, tree, xpath, namespaces, changed=False, msg="", hitcount=0, matches=[]):
```
Shouldn't use a mutable container as default arg in python.  If you ever add to the container you then find that you're sharing a single opy of that container everytime the function in python.  You can do one of these two instead:
``` python
# If you don't need to add to the container now.  This just prevents doing something stupid when the code is changed in the future.
 def finish(module, tree, xpath, namespaces, changed=False, msg="", hitcount=0, matches=tuple()): 

# if matches needs to be a mutable container.  This creates a fresh list everytime the function is called.
def finish(module, tree, xpath, namespaces, changed=False, msg="", hitcount=0, matches=None): 
    if matches is None:
        matches = []
```

abadger replied:
This applies to the other functions where a list is the default value as well.

abadger commented on lib/ansible/modules/files/xml.py:
```diff
@@ -0,0 +1,757 @@
+    elif in_type == 'yaml':
+        if isinstance(child, string_types):
+            return etree.Element(child)
+        elif isinstance(child, dict):
```
In general, it's better to use``` collections.MutableMapping``` in isinstance as that allows duck-typing.

dagwieers committed:
Various fixes after review

abadger commented on lib/ansible/modules/files/xml.py:
```diff
@@ -0,0 +1,757 @@
+
+def child_to_element(module, child, in_type):
+    if in_type == 'xml':
+        infile = BytesIO(child.encode('utf-8'))
```
Instead of child.encode(), use to_bytes(child, errors='surrogate_or_strict").  to_bytes() protect against calling encode() on something that is already a byte string. (calling encode on a byte string can sometimes traceback).

abadger commented on lib/ansible/modules/files/xml.py:
```diff
@@ -0,0 +1,757 @@
+    except Exception:
+        e = get_exception()
+        module.fail_json(msg="Xpath %s causes a failure: %s\n%s\n  -- tree is %s" %
+                             (xpath, e, traceback.format_exc(e), etree.tostring(tree, pretty_print=True)))
```
Instead of including a stacktrace in the message, better to put it into the exception field:
```
 module.fail_json(msg="Xpath %s causes a failure: %s\n  -- tree is %s" %
                             (xpath, to_native(e), etree.tostring(tree, pretty_print=True)),
                             exception=traceback.format_exc())
```

abadger commented on lib/ansible/modules/files/xml.py:
```diff
@@ -0,0 +1,755 @@
+try:
+    import json
+except ImportError:
+    import simplejson as json
```
Since we're now targetting python-2.6+ we can just ```import json```

abadger commented on lib/ansible/modules/files/xml.py:
```diff
@@ -0,0 +1,755 @@
+
+_re_splitSubLast = re.compile("^(.*)/(" + _nsIdent + ")\\[(.*)\\]$")
+
+_re_splitOnlyEqValue = re.compile("^(.*)/text\\(\\)=" + _xpstr + "$")
```
Unless there's a reason, I'd put all these toplevel variable definitions right after the imports.

They're also constants so they probably should be uppercase but I won't block on that.

abadger replied:
I see... it's putting them near split_xpath_last which is where they're used.  Maybe move both the variables and the split_xpath_last function up to below the imports.

abadger commented on lib/ansible/modules/files/xml.py:
```diff
@@ -0,0 +1,755 @@
+
+    m = _re_splitSubLast.match(xpath)
+    if m:
+        content = map(lambda x: x.strip(), m.group(3).split(" and "))
```
Another map that can be a list comprehension:
```
content = [x.strip() for x in m.group(3).split(" and ")]
```

dagwieers replied:
What's wrong with maps ?

abadger commented:
Okay, I'm done with my review once those last few things are fixed and shippable finally gets done testing the changes, I'll merge.  (If I don't notice that shippable is finished, feel free to ping me on IRC).

abadger commented:
Merged to devel for 2.4.0 release.  Thanks to @cmprescott @dagwieers and the many others who have worked on this until it could go in!

dagwieers commented:
Thanks @abadger !
